CLOUDSTACK-9298: Improve performance of resource retrieval that have tags associated and target volumes, VMs and templates

nvazquez opened the pull request:
JIRA TICKET: https://issues.apache.org/jira/browse/CLOUDSTACK-9298 

## Description of the problem
When retrieving a large number of resources which have tags associated with, retrieval methods took too long. Our goal is to improve performance of this methods avoiding query the database for each tag, managing that information in memory.
API methods to improve: <code>listTemplates</code>, <code>listVolumes</code>, <code>listVirtualMachines</code>

To achive it, it's necessary to include new columns in <code>template_view</code>, <code>volume_view</code> and <code>user_vm_view</code>:
* tag_account_name
* tag_domain_name
* tag_domain_uuid

rafaelweingartner commented:
@nvazquez , 
Lines 191-195 and 261-265 are duplicated, could you extract it to a method? And maybe even document it.

At the “ResourceTagJoinVO”, why did you create that new constructor?

Additionally, have you tested that view select? It looks ok, but I did not check clause per clause.

nvazquez commented:
Hi @rafaelweingartner,

Sure, I changed it to a method and added javadoc to it.

I added the new constructor at <code>ResourceTagJoinVO</code> to set all that fields in a single call instead of many setter methods. Which way could be better?

Yes, I tested <code>template_view</code> and it retrieves data.

Thanks,
Nicolas

pedro-martins commented:
@nvazquez, nice changes.

So, I guess that a test cases for these new methods will be important to detect future changes in these methods behavior. 

Another thing, is this new constructor really necessary? do you use this constructor in any part of the code? or do you just created this constructor to improve future codes calling only this constructor instead of lots of setters as you said? 

In my opinion, a constructor with many params is a bit ugly, I prefer to use no params in the constructor and use some factory pattern to create this object (lots of setters called inside a factory class), this is just my opinion.

Ty.

nvazquez commented:
Hi @pedro-martins thanks for your comments! 
About test cases you mean unit tests or an integration test for this changes?
I planned to use that constructor in similar methods which lists volumes, templates and vms, but I can remove it and follow your suggestion

pedro-martins commented:
@nvazquez 

I mean unit tests for each new method. 

I think that use the Builder to get an instance with all variables initialized is a good way.

Thanks.

rafaelweingartner commented:
@nvazquez and @pedro-martins, 
I am ok with the code now, expect the constructor bit.
The question you should ask about the constructor is, is it used NOW? If not, why create it. In the future when the need appear, we can create it; this way we can avoid unused code that may be left aside and never used.

Additionally, I am with @pedro-martins, if you like, tests cases would be pretty welcome here.

nvazquez commented:
@rafaelweingartner @pedro-martins thanks for your comments! I deleted the constructor and used setters instead and not included the Builder pattern, but if you want I can include it.
I also included unit tests for new methods.

pedro-martins commented on server/test/com/cloud/api/query/dao/TemplateJoinDaoImplTest.java:
```diff
@@ -0,0 +1,56 @@
+        assertEquals(0, templateResponse.getTags().size());
+        _templateJoinDaoImpl.updateTemplateTagInformation(template, templateResponse);
+        assertEquals(1, templateResponse.getTags().size());
+    }
```
Hi, @nvazquez.

I see a little mistake in you test, you are testing if the method has inserted a tag or not, I think that this test needs to verify if the template inserted in the templateResponse is the same that you has passed in the first param of the method.
You can do it checking if each variable in both templates are equals.

The same problem in the other tests cases.

Ty.

rafaelweingartner commented:
@nvazquez I have reviewed your code.
Class “VolumeJoinDaoImpl” at line 46, why did you change the get Logger method to use “VolumeJoinDaoImplTest”?

The method “updateTemplateTagInformation” is duplicated (I would say triplicated) at “VolumeJoinDaoImpl”, “UserVmJoinDaoImpl” and “TemplateJoinDaoImpl”. I know, you may argue that the method “updateTemplateTagInformation” has a conditional in “TemplateJoinDaoImpl” and “VolumeJoinDaoImpl”; still, that conditional is executed with some additional checks in “UserVmJoinDaoImpl”. The point here is that you should let that conditional out of the method scope, and then extract the method to a place that can be reused in all of those classes. All of them (VolumeJoinDaoImpl, UserVmJoinDaoImpl, and TemplateJoinDaoImpl) extend “GenericDaoBase” class. You can easily create a new class that extend “GenericDaoBase”, with that “updateTemplateTagInformation” and make all of those classes extend this new one. Please, keep that in mind in your next PRs.

That same is happening for VolumeJoinVO, UserVmJoinVO, and TemplateJoinVO classes. All of them have attributes “tagAccountName”, “tagDomainUuid” and “tagDomainName” and their respective getters and setters. They all extend “BaseViewVO”, you can do that same trick I mentioned you above here.

Now about the tests. First, when writing tests, you should ask yourself, “What am I going to test?”, “If this method changes the behavior would I like to catch it?”, “What is the behavior I want to monitor the behavior of?”. Those are the basic questions.
I believe your tests, should test the method “updateTemplateTagInformation” that you wrote. So, what do you want to detect if that method stops setting a property? Do you want to detect if it sets a property in a wrong place? If you want that, your tests now are not adequate, if you do not want that, you should delete the tests classes.

I personally believe that the method deserves tests, to check if it is working as expected. Meaning, that it sets the properties you want in the right place.

Additionally, when writing tests try to code as little as possible. Variables “accountService” and “configDao” are not needed to test the “updateTemplateTagInformation”; keep in mind that is a unit test. To test “updateTemplateTagInformation” you only need a simple instance of “TemplateJoinDaoImpl” and a mock of “ApiDBUtils.newResourceTagResponse(vtag, false)”

nvazquez commented:
@rafaelweingartner @pedro-martins I followed your suggestions and made a big refactor, hope this would be better now. I'm planning to run integration tests and post the results later.
Thanks!

rafaelweingartner commented on server/src/com/cloud/api/query/dao/GenericDaoBaseWithTagInformation.java:
```diff
@@ -0,0 +1,58 @@
+import com.cloud.api.query.vo.ResourceTagJoinVO;
+import com.cloud.utils.db.GenericDaoBase;
+
+public abstract class GenericDaoBaseWithTagInformation<T extends BaseViewWithTagInformationVO,
```
Why create a type ID that extends Serializable, if this type definition is always Long?
I think it is better to use "public abstract class GenericDaoBaseWithTagInformation<T extends BaseViewWithTagInformationVO, Z extends BaseResponseWithTagInformation> extends GenericDaoBase<T, Long> {..}"

nvazquez replied:
Cool, I changed it

rafaelweingartner commented on server/test/com/cloud/api/query/dao/GenericDaoBaseWithTagInformationBaseTest.java:
```diff
@@ -0,0 +1,59 @@
+        return tagResponse;
+    }
+
+    public void testUpdateTagInformation(GenericDaoBaseWithTagInformation<T, ID, Z> dao, T baseView, Z baseResponse,
```
What about checking every property that the method "updateTagInformation" touches?
Additionally, to test that, you do not need to pass as parameter "tagId, key and others" to set at the baseView. If you always use the same values (I see no problem on that), no need to receive those parameters.

nvazquez replied:
I added a new test case and changed values to base test, so there's need to pass them

pedro-martins commented:
@nvazquez.

Could you squash all of yours commits into one? It's hard to compare the original code with your current modifications. If you do not know how to squash here goes a [link](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/) with a little tutorial

nvazquez commented:
@pedro-martins Sure, I'll push some changes that @rafaelweingartner pointed out and squash them all into one

nvazquez commented:
Done @pedro-martins

nvazquez commented:
Hi @rafaelweingartner @pedro-martins,

I see continuous integration check is still failing due to this error, I made some changes in <code>setup/db/db/schema-481to490.sql</code>:

<code>[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:java (create-schema) on project cloud-developer: An exception occured while executing the Java class. null: InvocationTargetException: Unable to upgrade the database: Unable to execute upgrade script: /home/travis/build/apache/cloudstack/developer/target/db/db/schema-481to490.sql: Access denied; you need (at least one of) the SUPER privilege(s) for this operation -> [Help 1]</code>

rafaelweingartner commented:
@nvazquez the code is ok now.
I believe the problems you are having are due to lines 27, 28 and 29; I think you can remove them.

nvazquez commented:
Thanks @rafaelweingartner 

I removed those lines but still not working:

<code>[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:java (create-schema) on project cloud-developer: An exception occured while executing the Java class. null: InvocationTargetException: Unable to upgrade the database: Unable to execute upgrade script: /home/travis/build/apache/cloudstack/developer/target/db/db/schema-481to490.sql: Table 'user_vm_view' already exists -> [Help 1]</code>

rafaelweingartner commented:
Now the error changed.
It is in another table and probably is not related to your work.
Now the problem is with "user_vm_view"
Could you track who added that to the script?

On Fri, Mar 11, 2016 at 7:01 AM, Nicolas Vazquez <notifications@github.com>
wrote:

> Thanks @rafaelweingartner <https://github.com/rafaelweingartner>
>
> I removed those lines but still not working:
>
> [ERROR] Failed to execute goal
> org.codehaus.mojo:exec-maven-plugin:1.2.1:java (create-schema) on project
> cloud-developer: An exception occured while executing the Java class. null:
> InvocationTargetException: Unable to upgrade the database: Unable to
> execute upgrade script:
> /home/travis/build/apache/cloudstack/developer/target/db/db/schema-481to490.sql:
> Table 'user_vm_view' already exists -> [Help 1]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/cloudstack/pull/1425#issuecomment-195303532>.
>



-- 
Rafael Weingärtner

nvazquez commented:
@rafaelweingartner it was my mistake, I had writen user_vm__view instead of user_vm_view on the script. I fixed it and pushed again

rafaelweingartner commented:
@nvazquez now everything seems ok
Could you run integration tests?

nvazquez commented:
I post my test results for <code>test_vm_lifecycle.py</code>

````
[root@ussarlabcsmgt41 cloudstack]# nosetests --with-marvin --marvin-config=setup/dev/advanced.cfg test/integration/smoke/test_vm_life_cycle.py

==== Marvin Init Started ====

=== Marvin Parse Config Successful ===

=== Marvin Setting TestData Successful===

==== Log Folder Path: /tmp//MarvinLogs//Mar_11_2016_08_59_15_NX2IH0. All logs will be available here ====

=== Marvin Init Logging Successful===

==== Marvin Init Successful ====
===final results are now copied to: /tmp//MarvinLogs/test_vm_life_cycle_QS5H2K===
````
````
[root@ussarlabcsmgt41 cloudstack]# cat /tmp/MarvinLogs/test_vm_life_cycle_QS5H2K/results.txt
Test advanced zone virtual router ... === TestName: test_advZoneVirtualRouter | Status : SUCCESS ===
ok
Tests for basic zone virtual router ... === TestName: test_basicZoneVirtualRouter | Status : SUCCESS ===
ok
Test Deploy Virtual Machine ... === TestName: test_deploy_vm | Status : SUCCESS ===
ok
Test Multiple Deploy Virtual Machine ... === TestName: test_deploy_vm_multiple | Status : SUCCESS ===
ok
Test Stop Virtual Machine ... === TestName: test_01_stop_vm | Status : SUCCESS ===
ok
Test Start Virtual Machine ... === TestName: test_02_start_vm | Status : SUCCESS ===
ok
Test Reboot Virtual Machine ... === TestName: test_03_reboot_vm | Status : SUCCESS ===
ok
Test destroy Virtual Machine ... === TestName: test_06_destroy_vm | Status : SUCCESS ===
ok
Test recover Virtual Machine ... === TestName: test_07_restore_vm | Status : SUCCESS ===
ok
Test migrate VM ... === TestName: test_08_migrate_vm | Status : SUCCESS ===
ok
Test destroy(expunge) Virtual Machine ... === TestName: test_09_expunge_vm | Status : SUCCESS ===
ok
Test for attach and detach ISO to virtual machine ... === TestName: test_10_attachAndDetach_iso | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 12 tests in 858.458s

OK
````

rafaelweingartner commented:
@nvazquez thanks for running the functional tests...
I was wondering, is there any Marvin tests for templates and volumes (backup, creation and others)?
Because your changes also affect them.

nvazquez commented:
Hi @rafaelweingartner, I also ran <code>test/integration/smoke/test_templates.py</code>, I post its results:

````
[root@ussarlabcsmgt41 cloudstack]# cat /tmp/MarvinLogs/test_templates_LRM26P/results.txt
Test create public & private template ... === TestName: test_01_create_template | Status : SUCCESS ===
ok
Test when createTemplate is used to create templates having the same name all of them get ... === TestName: test_CreateTemplateWithDuplicateName | Status : SUCCESS ===
ok
Test Edit template ... === TestName: test_02_edit_template | Status : SUCCESS ===
ok
Test delete template ... === TestName: test_03_delete_template | Status : SUCCESS ===
ok
Test for extract template ... === TestName: test_04_extract_template | Status : SUCCESS ===
ok
Update & Test for template permissions ... === TestName: test_05_template_permissions | Status : SUCCESS ===
ok
Test for copy template from one zone to another ... SKIP: Not enough zones available to perform copy template
Test only public templates are visible to normal user ... === TestName: test_07_list_public_templates | Status : SUCCESS ===
ok
Test System templates are not visible to normal user ... === TestName: test_08_list_system_templates | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 9 tests in 5671.295s

OK
````

rafaelweingartner commented:
@nvazquez that is great. I can give an LGTM now.

nvazquez commented:
Thanks @rafaelweingartner!
@pedro-martins could you review the last changes?

pedro-martins commented:
@nvazquez, Nice changes, great job refactoring your implementation. I found no more troubles in this PR, then LGTM.

Ty.

serg38 commented:
Great!!! 2 LGTM received. Can we have this merged?

rafaelweingartner commented:
I guess so, if no one objects.

The code was extensively reviewed, test cases (unit ones) were written, functional test were executed. For me this PR is ready to be merged. I would be able to do that latter today.

swill commented:
Did we run the whole BVT test suite on it as well?  I see a few integration test run, but not many.  If we think it is ready to go, I can merge it in...

swill commented:
I am running tests against real hardware right now for this.  Assuming everything goes will I will merge it right after.  Thx...

rafaelweingartner commented:
ok thanks ;)

swill commented:
Testing my first PR with my shiny new CI environment.  :P

rafaelweingartner commented:
Out of curiosity, is it running automatically testing all of the opened PRs?
Curiosity 2, is it real hardware or a buble?

swill commented:
I have to manually kick off the PRs for now, I have not automated it yet.  Still getting going.

It is deployed as hypervisor on hypervisor using bubble, so it is running against a physical environment.  Does that answer your question?

rafaelweingartner commented:
Sure it answered ;)

swill commented:
I am hoping to get things more automated as I get more comfortable with everything, but I just had to get something in place so I can start getting through this PR backlog and start getting stuff merged in.

rafaelweingartner commented:
That is cool. This will certainly be a great benefit to our community.
What are the resources (CPU, memory, HD) requirements to run this appliance?

swill commented:
**Minimal**
CPU Intel 4 core 64-bit processor with virtualization extensions
RAM 16GB
HDD 20GB

**Recommended**
CPU Intel Skylake 4 core 64-bit processor with virtualization extensions
RAM 32GB
SSD 20GB (root) + 100GB (data)

SBP has kindly open sourced all of their work on this, so checkout the following links.  If you have questions, ping the dev list and I will try to make sure your questions are answered.

Setup an environment from scratch without an existing chef server:
https://github.com/MissionCriticalCloud/bubble-blueprint

The above link runs this cookbook:
https://github.com/MissionCriticalCloud/bubble-cookbook

This sets up the environment for you.  Once that is setup, you can use the following to configure ACS environments and run tests:
https://github.com/MissionCriticalCloud/bubble-toolkit

rafaelweingartner commented:
nice thanks

swill commented:
My tests verified that this was ready to merge.  Merged...

remibergsma commented:
@swill Congrats on the first bubble-tested PR :-) I would recommend posting the results though, the results.txt file shows which tests were run. If later we want to check, it's documented here.

swill commented:
Yes I meant to post the results, but I was too quick trying to get thing setup for the next run and I blew away my environment.  :(  Im a noob...  I will be posting the results going forward.
